File: FeatureDemos/Delphi/FireDACMetadata/Module.Main.pas
unit Module.Main;

interface

uses
  System.SysUtils, System.Classes, Web.HTTPApp, System.Generics.Collections, Web.Stencils,
  FireDAC.Stan.Intf, FireDAC.Stan.Option, FireDAC.Stan.Param,
  FireDAC.Stan.Error, FireDAC.DatS, FireDAC.Phys.Intf, FireDAC.DApt.Intf,
  Data.DB, FireDAC.Comp.DataSet, FireDAC.Comp.Client, FireDAC.Stan.Async,
  FireDAC.DApt, FireDAC.UI.Intf, FireDAC.Stan.Def, FireDAC.Stan.Pool,
  FireDAC.Phys, FireDAC.Phys.SQLite, FireDAC.Phys.SQLiteDef,
  FireDAC.Stan.ExprFuncs, FireDAC.Phys.SQLiteWrapper.Stat,
  FireDAC.ConsoleUI.Wait;

type

  TWebModuleMain = class(TWebModule)
    WSProcessor: TWebStencilsProcessor;
    [WebStencilsVar('customers', false)]
    Customers: TFDQuery;
    CustomersID: TFDAutoIncField;
    CustomersCOMPANY: TStringField;
    CustomersFIRST_NAME: TStringField;
    CustomersLAST_NAME: TStringField;
    CustomersGENDER: TStringField;
    CustomersAGE: TIntegerField;
    CustomersPOSTAL_CODE: TStringField;
    CustomersADDRESS: TStringField;
    CustomersCITY: TStringField;
    CustomersCOUNTRY: TStringField;
    CustomersPHONE: TStringField;
    CustomersEMAIL: TStringField;
    CustomersIP_ADDRESS: TStringField;
    CustomersACTIVATION_DATE: TDateField;
    CustomersACTIVE: TBooleanField;
    CustomersCOMMENTS: TWideMemoField;
    Connection: TFDConnection;
    procedure WebModuleMainDefaultHandlerAction(Sender: TObject;
      Request: TWebRequest; Response: TWebResponse; var Handled: Boolean);
    procedure WebModuleCreate(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  WebModuleClass: TComponentClass = TWebModuleMain;

implementation

{%CLASSGROUP 'System.Classes.TPersistent'}

{$R *.dfm}

procedure TWebModuleMain.WebModuleMainDefaultHandlerAction(Sender: TObject;
  Request: TWebRequest; Response: TWebResponse; var Handled: Boolean);
begin
  Customers.Locate('ID', Random(500), []);
  Response.Content := WSProcessor.Content;
end;

procedure TWebModuleMain.WebModuleCreate(Sender: TObject);
begin
  // Configure whitelist to expose FireDAC field metadata properties
  // This allows templates to access field properties like DisplayText, DataType, Required, etc.
  TWebStencilsProcessor.Whitelist.Configure(TField, ['DisplayText', 'Value', 'DisplayLabel', 'FieldName', 'Required', 'Visible', 'DataType', 'Size', 'IsNull'], nil, False);
  Connection.Params.Database := '..\..\database.sqlite3';
  customers.Active := True;
end;

end.

File: FeatureDemos/C++/FireDACMetadata/ModuleMain.cpp

//---------------------------------------------------------------------------
#include "ModuleMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"

TComponentClass WebModuleClass = __classid(TWebModuleMain);
//---------------------------------------------------------------------------

__fastcall TWebModuleMain::TWebModuleMain(TComponent* Owner)
  : TWebModule(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TWebModuleMain::WebModuleCreate(TObject *Sender)
{
  // Configure whitelist to expose FireDAC field metadata properties
  // This allows templates to access field properties like DisplayText, DataType, Required, etc.
  System::DynamicArray<System::UnicodeString> Props;
  Props.Length = 9;
  Props[0] = "DisplayText";
  Props[1] = "Value";
  Props[2] = "DisplayLabel";
  Props[3] = "FieldName";
  Props[4] = "Required";
  Props[5] = "Visible";
  Props[6] = "DataType";
  Props[7] = "Size";
  Props[8] = "IsNull";
  System::DynamicArray<System::UnicodeString> EmptyArray;
  TWebStencilsProcessor::Whitelist->Configure(__classid(TField), Props, EmptyArray, false);
  
  Connection->Params->Database = "..\\..\\database.sqlite3";
  Customers->Active = true;
  
  // Manually add the Customers query to WebStencils (C++ doesn't support attributes)
  WSProcessor->AddVar("customers", Customers, false);
}
//---------------------------------------------------------------------------

void __fastcall TWebModuleMain::WebModuleMainDefaultHandlerAction(TObject *Sender, TWebRequest *Request,
          TWebResponse *Response, bool &Handled)
{
  Customers->Locate("ID", Random(500), TLocateOptions());
  Response->Content = WSProcessor->Content();
}
//---------------------------------------------------------------------------

